### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by imanghafoori1**

### CODE REVIEW

File: src/TicketitServiceProvider.php [hunk 1]
```diff
@@ -55,42 +54,13 @@ public function boot()
             TicketItComposer::summerNotes();
 
             // Send notification when new comment is added
-            Comment::creating(function ($comment) {
-                if (Setting::grab('comment_notification')) {
-                    $notification = new NotificationsController();
-                    $notification->newComment($comment);
-                }
-            });
+            app(CommentObserver::class)->creating();
```
Comment (thekordy) [inline src/TicketitServiceProvider.php:57]: I think this should be `Comment::observe(CommentObserver::class);` ... see balping [comment](https://github.com/thekordy/ticketit/pull/491#issuecomment-416339807)

File: src/TicketitServiceProvider.php [hunk 2]
```diff
@@ -55,42 +54,13 @@ public function boot()
...
 
+            $ticketObserver = app(TicketObserver::class);
```
Comment (thekordy) [inline src/TicketitServiceProvider.php:59]: Same as above

File: src/TicketitServiceProvider.php [hunk 3]
```diff
@@ -55,42 +54,13 @@ public function boot()
...
             // Send notification when ticket status is modified
-            Ticket::created(function ($ticket) {
-                if (Setting::grab('assigned_notification')) {
-                    $notification = new NotificationsController();
-                    $notification->newTicketNotifyAgent($ticket);
-                }
-
-                return true;
-            });
+            $ticketObserver->created();
```
Comment (thekordy) [inline src/TicketitServiceProvider.php:63]: same as above

File: src/ModelsObservers/CommentObserver.php [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+use Kordy\Ticketit\Controllers\NotificationsController;
+use Kordy\Ticketit\Models\Comment;
+use Kordy\Ticketit\Models\Setting;
+
+class CommentObserver
+{
+    public function creating()
+    {
+        Comment::creating(function ($comment) {
```
Comment (thekordy) [inline src/ModelsObservers/CommentObserver.php:13]: I think we don't need to listen the event inside the observer, already done in the Service Provider, I believe we should remove that line, and get enough with the callback content.

May be:
```php
public function creating(Comment $comment)
{
    if (Setting::grab('comment_notification')) {
        $notification = new NotificationsController();
        $notification->newComment($comment);
     }
}
```

File: src/ModelsObservers/TicketObserver.php [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+use Kordy\Ticketit\Controllers\NotificationsController;
+use Kordy\Ticketit\Models\Setting;
+use Kordy\Ticketit\Models\Ticket;
+
+class TicketObserver
+{
+    public function created()
+    {
+        // Send notification when ticket status is modified
+        Ticket::created(function ($ticket) {
```
Comment (thekordy) [inline src/ModelsObservers/TicketObserver.php:14]: same as above

File: src/ModelsObservers/TicketObserver.php [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+                $notification->newTicketNotifyAgent($ticket);
+            }
+
+            return true;
+        });
+    }
+
+    public function updating()
+    {
+        Ticket::updating(function ($modified_ticket) {
```
Comment (thekordy) [inline src/ModelsObservers/TicketObserver.php:26]: same as above

**Comment by thekordy** [comment]
Good to think about refactoring the code, and great from you to share it :+1:  .. Thanks for your contributions.

**Review by thekordy — CHANGES_REQUESTED** [review]

File: src/TicketitServiceProvider.php (hunk 1 above)
Comment (imanghafoori1) [reply src/TicketitServiceProvider.php:57]: These classes I have created, have nothing to do with laravel model observers.
They are just normal php classes to put some code within them.
Actually I want my PR to be as understandable as possible for you. so you can accept them

**Comment by imanghafoori1** [comment]
I just wanted to put the code somewhere else and call it from the original place.
The main goal is to make the the whole package ready to have unit tests and feature tests.
And only then start to really refactor the code into shape.

**Comment by balping** [comment]
I agree that the service provider now is a mess, but how does extracting these make it more testable?

**Comment by imanghafoori1** [comment]
Since we are resolving the new classes through the IOC container (app() helper function) now we can mock them and replace them with mock objects.

`Setting::grab` should also become a Facade, to be mockable.

**Pull request closed without merging**